Expression node for visual shaders

Chaosus opened the pull request:
This is Expression node proposal - with this power you can create almost anything in Visual Shaders.
Closes #22243

![expression](https://user-images.githubusercontent.com/3036176/57582098-508a4c00-74c9-11e9-91ed-3346313fecbb.gif)

![expression3](https://user-images.githubusercontent.com/3036176/57582399-1753db00-74cd-11e9-88b9-ca6143601e10.gif)

You can add any number of input and output ports. You can also resize this node by dragging right bottom corner, if your expression is huge - otherwise it also have a scrollbar.

## Remarks

- The expression node(VisualShaderNodeExpression) is derived from new VisualShaderNodeGroupBase - the base class for this and the Groups/Comments (in future).

- Of course I checked the saving/deleting/removing/duplication for new type of node and its ports(this takes a ton of time). It may still contains some implicit bugs but overall it's stable.

- Currently, this node does not have an ability to rename - I think this ability should be added to GraphNode first. I prefered to not modify core components in this PR - only the parts for visual shader.

- Created output ports are initially setted to its default value(0.0 for scalar, vec3(0.0,0.0,0.0) or identity matrix for transform).

Chaosus committed:
Expression node for visual shaders

fire commented:
We talked in Discord.

1. There's problems with margins in the node.
2. Remove the build button and remove it with a count down timer.
```
08:41:08reduzjust make a timer
08:41:41reduzevery time the expression changes, start the time (so you dont let it timeout while user is typing) make it like 0.5 second
08:41:46reduzthen it timeouts, it applies it
08:41:53reduzI mean a Timer node
08:42:15reduzthis way it wont rebuild while typing
```

Norrox commented:
Nice!

Chaosus commented:
@iFire 

> There's problems with margins in the node.

Yeah, currently it's impossible, this requires to modify the GraphNode.. will work on this tomorrow, maybe it's required additional PR

> Remove the build button and remove it with a count down timer.

Fixed

fire commented:
Typing enter on the port input and port output does not confirm the creation of the port.

Chaosus commented:
> Typing enter on the port input and port output does not confirm the creation of the port.

Done, also setted default size for created expression node to (200, 150)

Chaosus commented:
Added syntax highlighting

![image](https://user-images.githubusercontent.com/3036176/57617389-5e9fa180-7589-11e9-9ec3-55b4a5eea764.png)

Chaosus commented:
@iFire I've added some kind of margin to it by modify the size of adjanced label(I didnt found the other way to do it sorry)

![expression4](https://user-images.githubusercontent.com/3036176/57636989-f5348880-75b2-11e9-882a-6499a739f915.gif)

Chaosus commented:
Get rid of update_graph which emitted every time the extension is changed(while still update the shader itself) - this should increase performance + removed some bugs..

gamergamesz commented:
Awesome!!!...Is there any way to key shader- input in animation graph

Chaosus commented:
@gamergamesz Its a subject for another issue, I think you can only apply these key to uniforms due to the nature of shaders.

gamergamesz commented:
I can't see any part of visual shader getting keyed, Spatial shader can get keyed so i thought there  might be a way to key visual shader too.

Chaosus commented:
> I can't see any part of visual shader getting keyed, Spatial shader can get keyed so i thought there might be a way to key visual shader too.

Ah, then its not implemented yet, I suggest to search for similar issue or create a new one - because this issue is not related to this - its an Expression node feature proposal

Chaosus commented:
Replaced backet button with new popup - where you can modify type and name of the port

![image](https://user-images.githubusercontent.com/3036176/57699775-29668280-7661-11e9-85bd-0d497f96aa04.png)

fire commented:
I wonder how:

[Vector] [cx] 
[Change Port] [Delete Port]

Looks.

What do you think?

Chaosus commented:
Like so ?

![image](https://user-images.githubusercontent.com/3036176/57701166-0e494200-7664-11e9-9204-cfee94b77533.png)

Megalomaniak commented:
Even better, yes. Now remove the button and change the port name label into a button that can be clicked to bring it up instead.

Calinou commented:
The expression code editor should use the monospace/"source" font, like the script and shader editors.

Chaosus commented:
@Calinou Like so ?
![image](https://user-images.githubusercontent.com/3036176/57701735-3eddab80-7665-11e9-810d-81e7bab5cd80.png)

Calinou commented:
@Chaosus Yes :slightly_frowning_face: 

I'm not sure if we should use the font size defined in the editor settings, or derive  a font resource with a slightly smaller size (like `size - 1`, which would make it `13` by default).

Chaosus commented:
Need to take a break, @Megalomaniak About your idea with button - Is there is any example of how it can be looks like ?

Megalomaniak commented:
Ideally a LineEdit would be the best according to _my own_ preference: Click on the port name and you can just change it's name, hit enter to apply name change. After clicking on the name as the LineEdit is editable/visible the floating panel with the port type selector and delete button would also appear, note that this way no field for name change would be needed in floating panel.

Not sure that the change port button should be needed either.

And that is what I was going to originally recommend, however I realized that making it a button might just be more inviting for a new user to press it and find out what can be done there.

Megalomaniak commented:
Also I think the plus buttons on either side could be renamed respectively add input and add output.

Megalomaniak commented:
Though honestly perhaps the inputs and outputs should simply be defined by the expression/snippet? As it is you are going to be introducing odd/new UI/UX designs here that are used nowhere else in godot AFAIK.

Chaosus commented:
I think, the Change button could be safely removed and any change in that LineEdit will rename the port accordingly, same for type. I dont think that clicking on label is worth enough to be implemented - it's not a frequent operation anyways.

> Also I think the plus buttons on either side could be renamed respectively add input and add output.

I prefer to leave them as plus buttons for compactibility + they are already have a tooltips

> Though honestly perhaps the inputs and outputs should simply be defined by the expression/snippet? As it is you are going to be introducing odd/new UI/UX designs here that are used nowhere else in godot AFAIK.

Well maybe I dont using other programms which using expressions.. 
The purpose is to simplify the code. If I do it in your way - the new users will need to write a list of input and output nodes inside expression(probably with "in" and "out" prefix keywords). Also by doing so I need to rewrite major part of the code which took a long time. I can do it only if @reduz will say yes.

Megalomaniak commented:
Yeah, I don't actually mind seeing some extra/new UI/UX conventions added and I think this case is fairly good, but if it is added it also creates a precedent to consider doing similar things elsewhere throughout the editor. Thus it's worth considering if it's really worth doing and how to best go about it. :)

Chaosus commented:
So the cons are

- More confusion to newcomers
- More text need to be written even for simple tasks
- Additional (and big) parser step

and pros are

- Can be copy-pasted to other expressions to immediatly create input and output ports
- Less clumpy UI by removing ports settings buttons
- Conformity with other programs where expressions are exist

fire commented:
I don't know if it is possible but I agree with @Megalomaniak too.

* Click add port and it'll add a left or right port with a placeholder. 
* Click on the port name and it is converted you can change it's name, hit enter to apply name change.
* The delete button is always available in the ui, note that this way no field for name change would be needed in floating panel.

In my next post will be a mockup.

See my original design.

![img](https://user-images.githubusercontent.com/32321/45783096-ad26b700-bc19-11e8-9ae7-21bd4773ccb1.png)

fire commented:
Here's a mockup.

![Expression Mockup](https://user-images.githubusercontent.com/32321/57718069-5f8c0e00-7631-11e9-8787-5cb2f015819b.png)

Megalomaniak commented:
@Chaosus Sounds about right, yeah.

The first two cons would be just cons for newcomers who have never written shaders by code, though for those already familiar with shader languages - I don't think they would even register as cons.

Additional pro(with it's own caveats) could be if the inputs are produced by shader uniforms so that you could just load someone elses shader as an expression. Could be context sensitive so when loaded into an expression node in fragment section/editor it would read/load the fragment part of that shader only? Cons would be that users unfamiliar might end up loading extremely large shaders, and overall the parser would likely have to be more complex too(so way more work I guess). I'm even inclined to consider it more of a con, but the positive would be that there would (almost?) automatically be a whole library of them to chose from for those who can't write their own. 

But yeah, I'm entirely fine with not following that path at all. @fire 's mockups are also looking pretty good to me.

Megalomaniak commented:
@fire How about you switch the places of trash-bin/delete button and the plus buttons in the second mock-up though? :)

I'm guessing the idea is that you could add new inputs in the middle of others, but it might be better to have a little widget per each for grabbing and rearranging them, so a new port can always be added just to the end then. Currently it seems like the delete button would only remove the last port in stack.

fire commented:
Due to feedback from iamchuthulu.

![Expression Mockup](https://user-images.githubusercontent.com/32321/57718545-5ea7ac00-7632-11e9-9c8b-4b95f68f41b5.png)

Megalomaniak commented:
Nice timing!

fire commented:
Cthulhu is giving feedback on the position of the + button. We'll be moving that and the trash can.

Chaosus commented:
Ok, then I need a several days to rethink and make the requested things correctly

Chaosus commented:
> Could be context sensitive so when loaded into an expression node in fragment section/editor it would read/load the fragment part of that shader only?

Well, maybe could, but I dont want to do that in this PR, probably later in another (its worked without it perfectly now)

Megalomaniak commented:
Yeah, no. That thought was only in the context of inputs and outputs being generated from the code. I don't expect you or anyone else to actually follow that one though. I find fire's most recent mock-up to be the easiest and likely the most optimal path to follow here.

Chaosus commented:
@fire so, if I do the stuff on your picture(lastest) you accept this PR as finished ?

fire commented:
* Button for add output and input is expanded for easibility
* It has label text.
* It is in the bottom for being in the same place all the time.

![Expression Mockup Candidate 5](https://user-images.githubusercontent.com/32321/57720012-d4f9dd80-7635-11e9-95b1-5546593ea0e2.png)

fire commented:
This is my final mockup, please comment and confirm by checkmark.

* Button for add output and input is expanded for easibility
* It has label text.
* It is in the bottom for being in the same place all the time.
* Trash nodes are not aligned by the add nodes.

![Expression Mockup Candidate 6](https://user-images.githubusercontent.com/32321/57720218-3e79ec00-7636-11e9-8501-12f31c95de1c.png)

Chaosus commented:
@Megalomaniak there could be an intermediate variant where this node will get a button to generate the ports from the code, while keeping the option to add ports manually

Chaosus commented:
@fire seems ok, it's just a matter of taste - but if you wish I make it so, without a problem

IAmActuallyCthulhu commented:
> @fire seems ok, it's just a matter of taste - but if you wish I make it so, without a problem

Myself and @TheDuriel wish it so also.

Megalomaniak commented:
If possible, an interesting idea could be to make it so modifier key(for example, alt) + left mouse button click-drag on any given node's port allows to drag and rearrange the respective ports order, this way no additional element is needed for initiating the dragging/rearrangement function. Nor any buttons for moving one up or down a stack.

Chaosus commented:
@IAmActuallyCthulhu of course, sorry. These changes are not hard to make, I think I can finish today..

Chaosus commented:
@IAmActuallyCthulhu, @fire

![image](https://user-images.githubusercontent.com/3036176/57757320-e27a9a80-76fd-11e9-9fa5-e431b46665ea.png)

Chaosus commented:
Ok, this can be a final view ?

![image](https://user-images.githubusercontent.com/3036176/57760638-a991f400-7704-11e9-88f8-7e2fe1fdcb18.png)

Chaosus commented:
I think maximum amount of characters in the port name should be limited for example to 12.. What do you think ?

fire commented:
You can remove the dialog box by generating the input / output name automatically. 

Not sure how to align the output and the view texture button, so I'll leave that as it is.

The other stuff looks good. Thanks.

fire commented:
I would make text box larger so when you click on the show texture button it can expand into the text area and have 2-5 lines of text shown without resizing the entire node.

fire commented:
This is probably a new pull request, but there is no margins for the texture preview.

![Screen Shot 2019-05-15 at 5 11 44 AM](https://user-images.githubusercontent.com/32321/57774655-f4454900-76cf-11e9-96a1-db32265d4592.png)

Chaosus commented:
I removed the add port popup, now ports will be created with default name "input"(index) or "output"(index), also added margins for preview, and fixed some bugs

fire commented:
I caused a crash while clicking on add outputs:

```
ERROR: remove: Condition ' p_elem->_root != this ' is true.
   At: ./core/self_list.h:83.
handle_crash: Program crashed with signal 11
Dumping the backtrace. Please include this when reporting the bug on https://github.com/godotengine/godot/issues
[1] 1   libsystem_platform.dylib            0x00007fff594a5b5d _sigtramp + 29
[2] 2   ???                                 0x00007f7fc5a1e80c 0x0 + 140186753296396
```

Not sure the exact steps to recreate.

Chaosus commented:
Yeah, I probably need to spend a few days more to fix these bugs.

fire commented:
From irc:

Created a new resource of visual shader. Added an expression node, clicked add output + 5 times and it crashes.

You can type the same variable name on the other side.


It's weird typing the name of the input, it compiles too quickly and stutters. Use the timer method?

Chaosus commented:
@fire sorry, my internet was broken for today, check it again for bugs - to avoid them I've added new rename popup - yes it's less ellegant than before but I think it's required

fire commented:
The new rename popup is less elegant, but it still achieves the goal of adding the ports before  naming them.

I was able to crash the visual shader expression by clicking on add outputs. Do you need better error reproduction steps?

fire commented:
```
Dumping the backtrace. Please include this when reporting the bug on https://github.com/godotengine/godot/issues
[1] 1   libsystem_platform.dylib            0x00007fff7442bb5d _sigtramp + 29
[2] 2   ???                                 0x0000002b00000000 0x0 + 184683593728
[3] HashMap<StringName, Color, HashMapHasherDefault, HashMapComparatorDefault<StringName>, (unsigned char)3, (unsigned char)8>::getptr(StringName const&) const (in godot.osx.tools.64) (hash_map.h:333)
[4] Control::get_color(StringName const&, StringName const&) const (in godot.osx.tools.64) (control.cpp:962)
[5] GraphNode::_notification(int) (in godot.osx.tools.64) (graph_node.cpp:214)
[6] GraphNode::_notificationv(int, bool) (in godot.osx.tools.64) (graph_node.h:38)
[7] Object::notification(int, bool) (in godot.osx.tools.64) (object.cpp:954)
[8] CanvasItem::_update_callback() (in godot.osx.tools.64) (canvas_item.cpp:455)
[9] MethodBind0::call(Object*, Variant const**, int, Variant::CallError&) (in godot.osx.tools.64) (method_bind.gen.inc:61)
[10] Object::call(StringName const&, Variant const**, int, Variant::CallError&) (in godot.osx.tools.64) (object.cpp:942)
[11] MessageQueue::_call_function(Object*, StringName const&, Variant const*, int, bool) (in godot.osx.tools.64) (message_queue.cpp:256)
[12] MessageQueue::flush() (in godot.osx.tools.64) (message_queue.cpp:305)
```

Chaosus commented:
Thats, probably a platform-specific bug, since I dont have any crash on Windows, and I see you are on OSX.. hmm.. I need to think. You get a crash simply by creating node and clicking few times on Add input or Add output port ?

fire commented:
It has to be a bare resource not attached to a scene.

Chaosus commented:
@fire try again

fire commented:
You don't seem to be able to reproduce the bug and it still occurs so here's a project.

[VS Examples.zip](https://github.com/godotengine/godot/files/3190077/VS.Examples.zip)

Add a expression node to new_visualshader_crash.tres. Open it in the file system view. Clicking on add input / add output enough times and it'll crash.

Chaosus commented:
![expression](https://user-images.githubusercontent.com/3036176/57906499-e257d780-7882-11e9-8b45-f720b05f974f.gif)

fire commented:
Here's a stack trace.

![Uploading Screen Shot 2019-05-16 at 11.24.36 PM.png…]()

fire commented:
Here's a stack trace.

![Screen Shot 2019-05-16 at 11 24 36 PM](https://user-images.githubusercontent.com/32321/57907304-e3ace400-7831-11e9-91d4-20e932129b9e.png)

fire commented:
Crashes here.

![Screen Shot 2019-05-16 at 11 26 33 PM](https://user-images.githubusercontent.com/32321/57907366-148d1900-7832-11e9-8ad2-3d8d765c9521.png)

fire commented:
```cpp
VisualShader::Type type = VisualShader::Type(edit_type->get_selected());
	graph->clear_connections();
	//erase all nodes
	for (int i = 0; i < graph->get_child_count(); i++) {

		if (Object::cast_to<GraphNode>(graph->get_child(i))) {
            graph->get_child(i)->queue_delete();
		}
	}
```

fire commented:
Trying this:

```
	//erase all nodes
	for (int i = 0; i < graph->get_child_count(); i++) {

		if (Object::cast_to<GraphNode>(graph->get_child(i))) {
            Node *node = graph->get_child(i);
            graph->remove_child(node);
            memdelete(node);
            i--;
		}
	}
```

Chaosus commented:
@fire I think I fixed it now - try again

fire commented:
The ui portion looks good, I can't type into the textbox though :(

fire reviewed (changes requested):
Let's push [change] label design to another issue or pull request, but other than that the expression node looks good and is ready to merge.

Edited:

To be clear, that means remove the ability to add the label.

![Screen Shot 2019-05-17 at 9 24 00 AM](https://user-images.githubusercontent.com/32321/57942179-ab36f580-7885-11e9-8650-b485e3fe1cab.png)

fire commented:
Reasoning behind removing custom name.

In other systems we use commenting to give context to nodes.

In your implementation we have a custom labels for expressions, if we want to do it, it should be done for all nodes, but that doesn't make sense. Why special case expression nodes then?

It needs further thought and I want the expression node merged :D

Chaosus commented:
Ok, I've removed it

fire reviewed (dismissed):
Can click 6 times before adding port triggers (clicking fast)

fire commented:
We're trying one more fix to make the output and input ports directly editable as a textfield.

Chaosus commented:
Done

fire commented:
We're trying 1.2s for the expression textbox, but it looks good.

fire commented:
Noticed this bug. 
![Screen Shot 2019-05-17 at 11 01 04 AM](https://user-images.githubusercontent.com/32321/57947188-1804bc80-7893-11e9-8c43-54a9de6097cc.png)

Chaosus commented:
Fixed

fire commented:
Dragging the  resize handle to the leftmost also breaks.

Chaosus commented:
Fixed

fire commented:
Mentioned in irc. Dragging the resize handle to the topmost also breaks.

Chaosus commented:
Fixed

Chaosus commented:
Want to rewrite a rename procedure a bit..

Chaosus commented:
@fire can you check it again for crashes please - I've changed the rename port procedure - now it will apply the changes only when focus is gone.

fire commented:
* No problems seen with rename
* Dragging the resizer is buggy. Same as before. May be lower priority since it only affects the top to bottom resize?

Chaosus commented:
Thats strange, I dont see any visual bugs on resize...

![vss](https://user-images.githubusercontent.com/3036176/58030970-2d464900-7b28-11e9-8856-3f6b0402c20f.gif)

fire commented:
Drag the resizer from the right to the other corner. There's two ways you
can do this.

On Mon, May 20, 2019 at 7:54 AM Yuri Roubinsky <notifications@github.com>
wrote:

> Thats strange, I dont see any visual bugs on resize...
>
> [image: vss]
> <https://user-images.githubusercontent.com/3036176/58030970-2d464900-7b28-11e9-8856-3f6b0402c20f.gif>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/28838?email_source=notifications&email_token=AAAH4QPHWP7KCDRALCZL5ALPWK3STA5CNFSM4HMKJ3C2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODVZCYAA#issuecomment-494021632>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAH4QPKYEKM65K5HWJTQ33PWK3STANCNFSM4HMKJ3CQ>
> .
>

fire commented:
[Screen Recording 2019-05-20 at 11.42.33 AM.zip](https://github.com/godotengine/godot/files/3199305/Screen.Recording.2019-05-20.at.11.42.33.AM.zip)

Here's a video.

Chaosus commented:
@fire I cannot reproduce it.. maybe you can do it yourself by creating a new PR(its non-critical) ? Otherwise its good to merge for now..

fire commented:
Sounds good to me.

akien-mga commented:
Thanks a lot!

Chaosus commented:
Finally, thank you and @fire !

Rabocop commented:
Hi! Thanks for your work. I noticed that there are problems with the input nodes. Could you check if is this normal or I'm doing something wrong (in order to open an issue or not)? Apparently doesn't make any sense. There is the video:

![bug3](https://user-images.githubusercontent.com/7796648/62411085-82054600-b5e6-11e9-994e-45690ab696f9.gif)

The thing is that I have two scalar inputs, but I can't use them together in the same vec3(input1,input2,0.0) for example unless I kinda "convert them" to floats.  As there's no code warnings I don't know what's happening.
Cheers!

Chaosus commented:
@Robocop Could you please post more high-quality gif ?

Chaosus commented:
@Rabocop Could you please post more high-quality gif ?

Chaosus commented:
Yeah, and if you open the output there should be an error - "Expected expression, found: SEMICOLON" ?

Rabocop commented:
Hi, sorry for the delay, this is the video.
[vsbug.zip](https://github.com/godotengine/godot/files/3475912/vsbug.zip)
Output says :  :72 - Unknown identifier in expression: yy
Cheers!

Chaosus commented:
Thanks ! Will do a bugfix soon

Chaosus commented:
@Rabocop ok, I fixed it - check it up ! ) and next time, please, create a separate issue

Rabocop commented:
> @Rabocop ok, I fixed it - check it up ! ) and next time, please, create a separate issue

Nice!! Next time I will open a separate issue. Thank you very much!
